h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -231,13 +230,12 @@ private void gridSearch(Grid<MP> grid) {
               e.printStackTrace(pw);
               Log.warn("Grid search: model builder for parameters " + params + " failed! Exception: ", e, sw.toString());
             }
-            
+            numberOfBuiltModels--;
             grid.appendFailedModelParameters(model != null ? model._key : null, params, e);
           }
         } catch (IllegalArgumentException e) {
           Log.warn("Grid search: construction of model parameters failed! Exception: ", e);
-          // Model parameters cannot be constructed for some reason
-          it.modelFailed(model);
+          numberOfBuiltModels--;
```

Comment:
this legacy `catch (IllegalArgumentException e)` looks so nasty with the nested `catch (RuntimeException e)` inside.
I'm sure its scope should have been limited to `params = it.nextModelParameters(model);`, avoiding you to have to handle `numberOfBuiltModels--` multiple times.
In worst case scenario — let's imagine  `grid.appendFailedModelParameters` throws an `IllegalArgumentException` — the counter will be decremented twice.

Reply:
@seb-h2o I just don't want to expand a scope of the PR.  I wanted to focus on the issue/fix that will let me to proceed with  https://github.com/h2oai/h2o-3/pull/4040 . I can create jira for refactoring this later.

Reply:
Please do, thanks.

Reply:
@seb-h2o I can put `numberOfBuiltModels--` as a last line after `appendFailedModelParameters` just for now.

Reply:
@seb-h2o https://0xdata.atlassian.net/browse/PUBDEV-7113